resource.c

```diff
@@ -0,0 +1,232 @@
+	 * information.
+	 */
+	res = (res_blk_t *)
+		malloc(sizeof(res_blk_t) + (sizeof(res_unit_t *)) * res_count);
+	res->res_count = res_count;
+
+	/* Allocate and set each resource information properly */
+	for (int i = 0; i < res_count; i++) {
+		/* Allocate memory for resource info */
+		temp = (res_unit_t *)malloc(sizeof(res_unit_t));
+		memset(temp, 0, sizeof(res_unit_t));
+		temp->status = RES_STATUS_EMPTY;
+
+		/* Some resource information are big and need extra allocation.
+		 * In these cases an address is returned which hold actual
+		 * resource information.
+		 */
+		switch (res_ids[i]) {
+		case RES_MEM_ACTIVE:
+		case RES_MEM_INACTIVE:
+		case RES_MEM_AVAILABLE:
+		case RES_MEM_FREE:
+		case RES_MEM_TOTAL:
+		case RES_MEM_PAGESIZE:
+		case RES_MEM_SWAPFREE:
+		case RES_MEM_SWAPTOTAL:
+		case RES_KERN_COMPILE_TIME:
+		case RES_KERN_RELEASE:
+		case RES_NET_ALLIFSTAT:
+			break;
+
+		case RES_NET_IFSTAT:
+			temp->data.ptr = (res_net_ifstat_t *)
+				malloc(sizeof(res_net_ifstat_t));
+			break;
+
+		case RES_MEM_INFOALL:
+			temp->data.ptr = (res_mem_infoall_t *)
+				malloc(sizeof(res_mem_infoall_t));
+			break;
+
+		default:
+			eprintf("Invalid resource ID: %d", res_ids[i]);
+			while (--i >= 0)
+				free(res->res_unit[i]);
+			free(temp);
+			free(res);
+			errno = EINVAL;
+			return NULL;
+		}
+
+		temp->res_id = res_ids[i];
+		res->res_unit[i] = temp;
+	}
+
+	return res;
+}
+
+/* Free resources allocated in res_build */
+void res_destroy_blk(res_blk_t *res)
+{
+	for (int i = 0; i < res->res_count; i++) {
+		/* Some resource information are big and needed extra
+		 * Free memory allocated for that.
+		 */
+		switch (res->res_unit[i]->res_id) {
+		case RES_NET_IFSTAT:
+		case RES_MEM_INFOALL:
+			free((res->res_unit[i])->data.ptr);
+			break;
+		}
+
+		/* Free memory tho hold this resource info */
+		free(res->res_unit[i]);
+	}
+	free(res);
+}
+
+/* read resource information corresponding to res_id, out should have been
+ * properly allocated by caller if required.
+ */
+int res_read(int res_id, void *out, void *hint, int pid, int flags)
+{
+	struct utsname t;
+
+	if (out == NULL) {
+		switch (res_id) {
+		/* In case of RES_NET_ALLIFSTAT memory is allocated on the
+		 * basis of number of interfaces available on system.
+		 * So out can be NULL in that case.
+		 */
+		case RES_NET_ALLIFSTAT:
+			break;
+
+		default:
+			eprintf("out argument cannot be NULL");
+			errno = EINVAL;
+			return -1;
+		}
+	}
+
+	/* Check if memory proc file is needed to open */
+	if (res_id >= MEM_MIN && res_id < MEM_MAX)
+		return getmeminfo(res_id, out, hint, pid, flags);
+
+	/* Check if net proc file is needed to open */
+	if (res_id >= NET_MIN && res_id < NET_MAX)
+		return getnetinfo(res_id, out, hint, pid, flags);
+
+	switch (res_id) {
+	case RES_KERN_RELEASE:
+		uname(&t);
+		strcpy(out, t.release);
+		break;
+
+	case RES_KERN_COMPILE_TIME:
+		uname(&t);
+		sscanf(t.version, "%*s%*s%*s%[^\t\n]", (char *) out);
```

Comment:
I'm not sure whether this is some nifty use of a glibc feature or what, but the manpage for sscanf says that having more % specifiers than target pointers results in undefined behavior.